odl/deform/linearized.py

```diff
@@ -0,0 +1,545 @@
+# Copyright 2014-2016 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Operators and functions for linearized deformation."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+
+from odl.operator.operator import Operator
+from odl import ProductSpace, DiscreteLp, Gradient, PointwiseInner, Divergence
+import numpy as np
+
+
+__all__ = ('LinDeformFixedTempl', 'LinDeformFixedTemplDeriv',
+           'LinDeformFixedTemplDerivAdj', 'LinDeformFixedDisp',
+           'LinDeformFixedDispAdj', 'linear_deform', 'deform_grad')
+
+
+def linear_deform(template, displacement):
```

Comment:
We need to make a decision if this function is supposed to be public or not. Here are the consequences for the two scenarios:

**Public:**

- The name can stay as it is.
- The interface needs to be more flexible. We must accept input that is not yet in any space, for both `template` and `displacement`.
- We probably have to add a mandatory `space` argument, so that we can do `template = space.element(template)`, `displ_space = ProductSpace(space, space.ndim)` and `displacement = displ_space.element(displacement)`.

This is doable and looks more like a function that is useful for other people.

**Private:**

- We change the name to `_linear_deform`
- We keep the interface, making it a function only for internal use.
 
In the long run, we probably want this a public function, but the second variant would be okay for me, too. Just the current state is somewhere in between, and this needs to be resolved.